Refactor OSv compiler

miha-plesko opened the pull request:
OSv compiler needed some changes to match Rumprun compiler structure better. Prior this change, OSv compiler had a reversed logic compared to Rumprun. Fixed.

miha-plesko commented:
Mr @ilackarms , Mr @gberginc, as suggested in other [PR](https://github.com/emc-advanced-dev/unik/pull/110) the OSv compiler refactoring is now sitting in this separate PR. I kindly ask for a review.

miha-plesko commented:
I'd suggest to review this PR by examining directory structure and files inside `pkg/compilers/osv`, not the diff. Diff is ugly since I've moved some files and reorganized some functions, but the result is really neat to read IMHO.

miha-plesko commented:
Hello, does this refactoring seem reasonable to you guys? 😄

gberginc commented:
Hey, sorry @miha-plesko. I didn't know who you were addressing with Mr :). I'll take a look.

Commit:
Refactor OSv compiler without changing direcotry structure

OSv compiler needed some changes to match Rumprun compiler
structure better. Prior this change, OSv compiler had a reversed
logic compared to Rumprun. Fixed.

Signed-off-by: Miha Pleško <miha.plesko@xlab.si>

miha-plesko commented:
I'd be happy to see Capstan integration converging towards merge, if possible. I've rebased the PR to the latest master and dropped changing directory structure to simplify the review.

@ilackarms, @gberginc, gentlemen, I would kindly ask you for a review.

gberginc commented on pkg/compilers/osv/generic.go:
```diff
@@ -4,6 +4,35 @@ import (
+	CompileParams types.CompileImageParams
+}
+
+// CompilerHelper implements conversion of Capstan result into provider-like image.
```
provider-like --> provider-specific or something similar.

miha-plesko replied:
Done.

gberginc commented on pkg/compilers/osv/generic.go:
```diff
@@ -4,6 +4,35 @@ import (
+	UseEc2() bool
+}
+
+// CompilerHelperBase should be embedded in every CompilerHelper implementation.
```
Probably only in those that don't use EC2, right? See [below](https://github.com/emc-advanced-dev/unik/pull/111/files#diff-875d439e99d3786e7999650fc7b634f5R14).

miha-plesko replied:
I'd say that everyone should embed it and then override `UseEc2()` if needed. Isn't it more readable this way?

gberginc commented on pkg/compilers/osv/osv_aws.go:
```diff
@@ -1,23 +1,36 @@
-func (osvCompiler *OsvAwsCompiler) CompileRawImage(params types.CompileImageParams) (_ *types.RawImage, err error) {
-	resultFile, err := osvCompiler.CreateImage(params, true)
+type AwsCompilerHelper struct {
+	CompilerHelperBase
```
Why do you need to embed `CompilerHelperBase`? `AwsCompilerHelper` implements `UseEc2` function so there is no point in embedding it.

miha-plesko replied:
For the sake of pattern. What if we add another function to CompilerHelperBase later? I'd leave it be unless you insist.

gberginc commented on pkg/compilers/osv/osv_aws.go:
```diff
@@ -1,23 +1,36 @@
+}
+
+func (b *AwsCompilerHelper) Convert(params ConvertParams) (*types.RawImage, error) {
+	// Convert to WMDK format.
```
VMDK

gberginc commented on pkg/compilers/osv/osv_vmware.go:
```diff
@@ -1,23 +1,36 @@
+}
+
+func (b *VmwareCompilerHelper) Convert(params ConvertParams) (*types.RawImage, error) {
+	// Convert to WMDK format.
```
VMDK

gberginc commented on pkg/compilers/osv/osv_aws.go:
```diff
@@ -1,23 +1,36 @@
+
+func (b *AwsCompilerHelper) Convert(params ConvertParams) (*types.RawImage, error) {
+	// Convert to WMDK format.
+	resultFile, err := ioutil.TempFile("", "osv-boot.vmdk.")
```
This conversion is repeated in all helpers. I suggest you add it to the base helper (if so, ignore the embed comment above). Alternatively, you can add a utility function because the helper is not used in this scope (lines 18-30 in AWS).

gberginc replied:
As we discussed in Slack, it's also not perfectly clear why this "conversion" is strictly necessary - it basically just renames the file.

gberginc reviewed (commented):
@miha-plesko I like the changes. Even though they may not seem important, they help understanding what is going on and separate the task of  compiling the application and creating the image. Even though in OSv compilation actually means building the image, the task of composing image (with Capstan) is now nicely separated from task of creating and defining provider-specific image.

Nevertheless, I do have some comments trying to reduce the repetition of code, in particular the one referring to "converting" the image.

@ilackarms do you see any obvious mishaps that a change like this would break? This would simplify introduction of new compilers.

miha-plesko reviewed (commented):
Thank you @gberginc, I've applied your first-pass comments. Most importantly, I've dropped unnecessary file renames into .vmdk. Tested on OpenStack and VirtualBox and everything seems to be working without those. Thanks for noticing.
